### ISSUE TITLE
Spring boot dedicated management context causes Tooglz warning during startup

### ISSUE THREAD

**Issue by jpmsilva** [togglz/togglz#279]
With Togglz 2.5.0.Final and Spring Boot, if we setup the actuator endpoints in a dedicated context (by using the property `management.server.port` to specify a separate port), Spring will emit two `ContextRefreshedEvent` events, one for the main context and another for the management context.
This results in the following warning by Tooglz in the log:

`2018-06-29 13:52:41.414  WARN 29069 --- [           main] licationContextBinderApplicationListener : ApplicationContext already bound to current context class loader, releasing it first`

I have placed a breakpoint in `TogglzApplicationContextBinderApplicationListener.onApplicationEvent` and run several tests, and it appears that the management event is always received first.
This means that after the application starts up, the `ContextClassLoaderApplicationContextHolder` ends up containing the correct (main) application context.
However, I cannot be certain that the order of these events is always the observed one, and I don't think Spring Boot makes no such guarantees, which means that even if the order is correct now, it may not be in some future version.

I think that Togglz should not rely on the order of these events, and I propose that `ContextRefreshedEvent` events originating from the management context should instead be ignored.

If you agree with this idea, I can work on a PR to implement it.

Thanks and best regards

**Comment by chkal** [comment]
Thanks for reporting this!

Could you perhaps give `2.6.0-SNAPSHOT` a try? I was actually planing to release `2.6.0.Final` this weekend. And @ruifigueira spent quite some time reworking the Spring Boot support. Especially for the new Spring Boot 2.0 support.

@ruifigueira Maybe we should fix this before pushing out 2.6.0?

**Comment by jpmsilva** [comment]
Hi @chkal 

Just tested with a fresh build of 2.6.0-SNAPSHOT from the master branch, and I have the exact same result:

`2018-06-29 17:13:42.292  WARN 13826 --- [           main] licationContextBinderApplicationListener : ApplicationContext already bound to current context class loader, releasing it first`

I don't know if @ruifigueira will be able to reply soon (we actually work together). I'll ask him if he has some spare time to look at this.

**Comment by jpmsilva** [comment]
I took a quick look at the code and I have no good news :disappointed: 

The problem I have is that identifying which context a ContextRefreshedEvent belongs to depends on casting the source of the event to a `org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext`, and comparing it's `serverNamespace` to the value `management`.
This class is actually from Spring Boot. Given that the module where `TogglzApplicationContextBinderApplicationListener` resides,  `tooglz-spring-core` compiles against Spring proper (not Spring Boot), I don't really see an easy and elegant way out of this particular issue.
I could just compare the class names (instead of `instanceof`) and use reflection to call `getServerNamespace`, which is every bit of dirty as it sounds, but it would get the job done. Given that `ContextRefreshedEvent` events are rare during production usage, I would expect the performance impact of using reflection to be negligible.

Do you have any further ideas or comments?

**Comment by chkal** [comment]
I guess @ruifigueira can share some thoughts about this.

For me it is most important to know if we consider this a blocker for the release of 2.6.0.Final. And my current impression is that it is **not** a blocker. I absolutely agree that we should fix this, but many people are waiting for proper Spring Boot 2.0 support. So I would prefer to ship 2.6.0.Final this weekend and include the fix for this issue in 2.6.1.Final.

Let me know what you think.

**Comment by jpmsilva** [comment]
Well, given that this behavior has not been introduced with the Spring Boot 2 support, and we have been living with it in the previous releases, I would not consider it a blocker for the 2.6.0 release.

I will discuss options with @ruifigueira as well, and keep this on the backburner.

Thanks for the prompt feedback!

**Comment by chkal** [comment]
And thank you for bringing this up and your offer to help.... 🍻

**Comment by chkal** [comment]
Sorry, I forgot to mention that AFAIK @ruifigueira is on vacation, so it may take some time to get a response...

**Comment by jpmsilva** [comment]
No problem, I work with @ruifigueira (same company, same project) so I'm aware 😄 

I created a tentative fix in the form of PR #280 but I think it is perfectly fine to wait for 2.6.1.Final to consider any fixes for this issue.

Have a nice weekend!

**Comment by ruifigueira** [comment]
I'm officially back from vacations :) @jpmsilva, I took a look on your PR. I'll chat with you today to discuss this further.

@chkal , just to let you know that @jpmsilva coding skills are like gospel to me, I work with him and I'm very pleased he decided to take some of his time helping us 👍

**Comment by jpmsilva** [comment]
@chkal related to upgrading the minimum Java version to 8, it appears that the plugin org.eclipse.virgo.bundlor.maven that is being used to generate the manifest files cannot properly process Java classes containing lambdas.
It appears that this plugin is no longer being maintained.
I tried doing a drop-in replacement with bnd-maven-plugin, but I could not easily resolve the issues introduced by the different configuration styles and options.

Do you have some thoughts on this?
Thanks!

**Comment by jpmsilva** [comment]
It completely fails the build.
I don't know if tweaking the plugin configuration `<failOnWarnings>true</failOnWarnings>` would have any result, but I imagine that failure to process class bytecode will result in incomplete Import-Package and Export-Package directives, since these appear to be automatically generated by Virgo Bundlor - see 
[7.1.1 Export Package](http://www.eclipse.org/virgo/documentation/bundlor-documentation-1.1.2.RELEASE/docs/user-guide/htmlsingle/user-guide.html#d0e1246) and [7.1.2 Import Package](http://www.eclipse.org/virgo/documentation/bundlor-documentation-1.1.2.RELEASE/docs/user-guide/htmlsingle/user-guide.html#d0e1251).

I was reviewing the whole code base to look for possible places where verbose constructs (such as anonymous Runnables, Predicates, etc...) could be turned into lambdas, but as soon as I started, the build stopped working.

I think sooner or later the fact that Virgo Bundlor is no longer maintained will catch up with Togglz.
Do you know how we could involve OSGi people using Togglz - other than removing support for it and waiting for people to complain :smile:

**Comment by chkal** [comment]
Agreed. Not being able to use lambdas is really bad. We should really get rid of Bundlor.

Could you create a separate issue for that?

Perhaps we can lookup who asked for and who contributed the OSGi support in the issue tracker.

**Comment by chkal** [comment]
@jpmsilva Thanks for creating #290.
